File: crates/krishiv-python/examples/batch_delta_audit.py
#!/usr/bin/env python3
"""Delta Lake time-travel query and audit batch example in Python."""

import os
import json
import time
import tempfile
import pyarrow as pa
import pyarrow.parquet as pq
import krishiv as ks

def write_delta(root_path, table, mode="append"):
    """Helper to write local Delta Lake commits (Parquet + JSON metadata)."""
    log_dir = os.path.join(root_path, "_delta_log")
    os.makedirs(log_dir, exist_ok=True)
    
    # Get next version number
    max_ver = -1
    for f in os.listdir(log_dir):
        if f.endswith(".json"):
            try:
                v = int(f.split(".")[0])
                if v > max_ver:
                    max_ver = v
            except ValueError:
                pass
    version = max_ver + 1
    
    # Overwrite processing
    removed_files = []
    if mode == "overwrite":
        for f in os.listdir(root_path):
            if f.endswith(".parquet"):
                fp = os.path.join(root_path, f)
                removed_files.append(f)
                try:
                    os.remove(fp)
                except OSError:
                    pass
                    
    # Write data parquet file
    file_name = f"part-{version:05d}.parquet"
    file_path = os.path.join(root_path, file_name)
    pq.write_table(table, file_path)
    file_size = os.path.getsize(file_path)
    
    # Write delta commit JSON
    commit_info = {"commitInfo": {"operation": "WRITE", "timestamp": int(time.time() * 1000)}}
    add_info = {"add": {"path": file_name, "size": file_size, "dataChange": True}}
    
    log_path = os.path.join(log_dir, f"{version:020d}.json")
    with open(log_path, "w") as f:
        f.write(json.dumps(commit_info) + "\n")
        f.write(json.dumps(add_info) + "\n")
        for removed in removed_files:
            f.write(json.dumps({"remove": {"path": removed, "dataChange": True}}) + "\n")

def main():
    with tempfile.TemporaryDirectory() as temp_dir:
        delta_path = os.path.join(temp_dir, "my_delta_table")
        os.makedirs(delta_path, exist_ok=True)

        # 1. Create a Delta Lake table and write Version 0
        schema = pa.schema([
            ("id", pa.int64()),
            ("name", pa.string()),
        ])
        table_v0 = pa.Table.from_pydict({
            "id": [1, 2],
            "name": ["Alice", "Bob"],
        }, schema=schema)
        
        write_delta(delta_path, table_v0, mode="overwrite")

        # 2. Append to the table to create Version 1
        table_v1 = pa.Table.from_pydict({
            "id": [3],
            "name": ["Charlie"],
        }, schema=schema)
        
        write_delta(delta_path, table_v1, mode="append")

        # 3. Build the embedded session
        session = ks.Session.from_env()

        # 4. Query the latest version (Version 1)
        print("--- Current Version (Latest) ---")
        current_df = ks.read_delta(session, delta_path, version=None)
        print(current_df.collect().pretty())

        # 5. Query the historical version 0 (Time Travel!)
        print("--- Historical Version 0 ---")
        historical_df = ks.read_delta(session, delta_path, version=0)
        print(historical_df.collect().pretty())

if __name__ == "__main__":
    main()

File: crates/krishiv/examples/batch_delta_audit.rs
//! Delta Lake time-travel query and audit batch example.
//! Run with: `cargo run -p krishiv --example batch_delta_audit`

#![forbid(unsafe_code)]

use std::error::Error;
use std::sync::Arc;

use arrow::array::{Int64Array, StringArray};
use krishiv::{DataType, ExecutionMode, Field, RecordBatch, Schema, Session};
use tempfile::tempdir;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let temp = tempdir()?;
    let delta_path = temp.path().to_string_lossy().to_string();

    // 1. Create a Delta Lake table and write Version 0
    let schema = Arc::new(Schema::new(vec![
        Field::new("id", DataType::Int64, false),
        Field::new("name", DataType::Utf8, false),
    ]));
    let batch_v0 = RecordBatch::try_new(
        schema.clone(),
        vec![
            Arc::new(Int64Array::from(vec![1, 2])),
            Arc::new(StringArray::from(vec!["Alice", "Bob"])),
        ],
    )?;
    krishiv_lakehouse::write_delta(
        &delta_path,
        vec![batch_v0],
        krishiv_lakehouse::DeltaWriteMode::Overwrite,
        false,
    )
    .await?;

    // 2. Append to the table to create Version 1
    let batch_v1 = RecordBatch::try_new(
        schema.clone(),
        vec![
            Arc::new(Int64Array::from(vec![3])),
            Arc::new(StringArray::from(vec!["Charlie"])),
        ],
    )?;
    krishiv_lakehouse::write_delta(
        &delta_path,
        vec![batch_v1],
        krishiv_lakehouse::DeltaWriteMode::Append,
        false,
    )
    .await?;

    // 3. Build the embedded session
    let session = Session::builder()
        .with_execution_mode(ExecutionMode::Embedded)
        .build()?;

    // 4. Query the latest version (Version 1)
    println!("--- Current Version (Latest) ---");
    let current_df = session.read_delta_async(&delta_path, None).await?;
    println!("{}", current_df.collect_async().await?.pretty()?);

    // 5. Query the historical version 0 (Time Travel!)
    println!("--- Historical Version 0 ---");
    let historical_df = session.read_delta_async(&delta_path, Some(0)).await?;
    println!("{}", historical_df.collect_async().await?.pretty()?);

    Ok(())
}
